internal/api.go

```diff
@@ -320,31 +326,15 @@ func BackgroundContext() netcontext.Context {
 // any API calls are sent to the provided URL. It returns a closure to delete
 // the registration.
 // It should only be used by aetest package.
-func RegisterTestRequest(req *http.Request, apiURL *url.URL, decorate func(netcontext.Context) netcontext.Context) func() {
+func RegisterTestRequest(req *http.Request, apiURL *url.URL, decorate func(netcontext.Context) netcontext.Context) (*http.Request, func()) {
 	c := &context{
 		req:    req,
 		apiURL: apiURL,
 	}
-	ctxs.Lock()
-	defer ctxs.Unlock()
-	if _, ok := ctxs.m[req]; ok {
-		log.Panic("req already associated with context")
-	}
-	if _, ok := ctxs.dec[req]; ok {
-		log.Panic("req already associated with context")
-	}
-	if ctxs.dec == nil {
-		ctxs.dec = make(map[*http.Request]func(netcontext.Context) netcontext.Context)
-	}
-	ctxs.m[req] = c
-	ctxs.dec[req] = decorate
-
-	return func() {
-		ctxs.Lock()
-		delete(ctxs.m, req)
-		delete(ctxs.dec, req)
-		ctxs.Unlock()
-	}
+	ctx := withContext(decorate(req.Context()), c)
```

Comment:
Seems like you could factor out this pattern of `c.req = req.WithContext(withContext(decorate(req.Context()), c)`.  It comes up in the `handleHTTP` function, too.

Reply:
Is that easier to read?

I did it this way to increase readability. I might even lean toward unfactoring it even more.

Reply:
It had to stop and stare at it for a while to understand what's going on.  It's quite subtle, and I think part of what's tripping me up is the reassignment.  I think un-inlining is good.  What about:

```go
c := &context{apiURL: apiURL}
ctx := withContext(decorate(req.Context()), c)
req = req.WithContext(ctx)
c.req = req
return req, func() {}
```

Reply:
Doesn't work, because withContext looks at c.req.

Reply:
Never mind.  This is more of a rathole than I expected.